src/playground/editor.ts

```diff
@@ -0,0 +1,118 @@
+import { loadWASM } from "onigasm";
+import { Registry } from "monaco-textmate";
+import { wireTmGrammars } from "monaco-editor-textmate";
+import grainGrammar from "../../grain-language-server/editor-extensions/vscode/syntaxes/grain.json";
+import darkTheme from "../../themes/github-dark-modified-converted.json";
+import lightTheme from "../../themes/github-light-converted.json";
+import onigasmPath from "onigasm/lib/onigasm.wasm?url";
+import EditorWorker from "monaco-editor/esm/vs/editor/editor.worker?worker";
+
+const languageConfiguration: monaco.languages.LanguageConfiguration = {
+  "comments": {
+    "lineComment": "//",
+    "blockComment": ["/*", "*/"],
+  },
+  // symbols used as brackets
+  "brackets": [
+    ["{", "}"],
+    ["[", "]"],
+    ["(", ")"],
+  ],
+  // symbols that are auto closed when typing
+  "autoClosingPairs": [
+    { open: "{", close: "}" },
+    { open: "[", close: "]" },
+    { open: "(", close: ")" },
+    { open: "\"", close: "\"" },
+    { open: "'", close: "'" },
+    { open: "/*", close: "*/" }
+  ],
+  // symbols that that can be used to surround a selection
+  "surroundingPairs": [
+    { open: "{", close: "}" },
+    { open: "[", close: "]" },
+    { open: "(", close: ")" },
+    { open: "\"", close: "\"" },
+    { open: "'", close: "'" }
+  ],
+  // Based on https://github.com/kevb34ns/auto-comment-blocks/blob/24f8ebc584e3e77d6a116a5a03a80c55008b44a3/src/rules.ts#L7-L40
+  "onEnterRules": [
+    {
+      // e.g. /** | */
+      "beforeText": new RegExp("^\\s*\\/\\*\\*(?!\\/)([^\\*]|\\*(?!\\/))*$"),
+      "afterText": new RegExp("^\\s*\\*\\/$"),
+      "action": { "indentAction": monaco.languages.IndentAction.IndentOutdent, "appendText": " * " }
+    },
+    {
+      // e.g. /** ...|
+      "beforeText": new RegExp("^\\s*\\/\\*\\*(?!\\/)([^\\*]|\\*(?!\\/))*$"),
+      "action": { "indentAction": monaco.languages.IndentAction.None, "appendText": " * " }
+    },
+    {
+      // e.g.  * ...|
+      "beforeText": new RegExp("^(\\t|(\\ ))*\\ \\*(\\ ([^\\*]|\\*(?!\\/))*)?$"),
+      "action": { "indentAction": monaco.languages.IndentAction.None, "appendText": "* " }
+    },
+    {
+      // e.g.  */|
+      "beforeText": new RegExp("^(\\t|(\\ ))*\\ \\*\\/\\s*$"),
+      "action": { "indentAction": monaco.languages.IndentAction.None, "removeText": 1 }
+    }
+  ]
+}
+
+window.MonacoEnvironment = {
+  getWorker(moduleId, label) {
+    switch (label) {
+      case "editorWorkerService":
+        return new EditorWorker();
+    }
+    throw new Error(`Unimplemented worker ${label} (${moduleId})`);
+  },
+}
+
+await loadWASM(onigasmPath);
+
+const registry = new Registry({
+  getGrammarDefinition: async () => {
+    return {
+      format: 'json',
+      content: grainGrammar,
+    }
+  }
+});
+
+const grammars = new Map([
+  ['grain', 'source.grain'],
+]);
+
+export async function createGrainEditor(id: string, value: string) {
```

Comment:
I think we should open an issue to use [codemirror](https://codemirror.net/) on mobile, monaco has never really worked great on mobile and is a rather large import.

Reply:
Interesting, I just tried on both a Pixel 8 (Android 14, Chrome 131) and an iPhone 15 pro (iOS 18, Safari) and basic editing functionality works fine. Did you experience problems on your phone?

Reply:
https://github.com/microsoft/monaco-editor/issues/246 discusses a lot of the issues, while it's come a long way and a lot of the editing works fine I know there are some edge cases related to selection and typing. Currently on mobile I experience some small highlighting issues (nothing major), but I think it would be worth opening an issue to address it in the future.